src/ripple/app/paths/impl/AMMLiquidity.cpp

```diff
@@ -0,0 +1,205 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+#include <ripple/app/paths/AMMLiquidity.h>
+#include <ripple/app/paths/AMMOffer.h>
+
+namespace ripple {
+
+template <typename TIn, typename TOut>
+AMMLiquidity<TIn, TOut>::AMMLiquidity(
+    ReadView const& view,
+    AccountID const& ammAccountID,
+    std::uint32_t tradingFee,
+    Issue const& in,
+    Issue const& out,
+    AMMContext& ammContext,
+    beast::Journal j)
+    : ammContext_(ammContext)
+    , ammAccountID_(ammAccountID)
+    , tradingFee_(tradingFee)
+    , issueIn_(in)
+    , issueOut_(out)
+    , initialBalances_{fetchBalances(view)}
+    , j_(j)
+{
+}
+
+template <typename TIn, typename TOut>
+TAmounts<TIn, TOut>
+AMMLiquidity<TIn, TOut>::fetchBalances(ReadView const& view) const
+{
+    auto const assetIn = ammAccountHolds(view, ammAccountID_, issueIn_);
+    auto const assetOut = ammAccountHolds(view, ammAccountID_, issueOut_);
+    // This should not happen.
+    if (assetIn < beast::zero || assetOut < beast::zero)
+        Throw<std::runtime_error>("AMMLiquidity: invalid balances");
+
+    return TAmounts{get<TIn>(assetIn), get<TOut>(assetOut)};
+}
+
+template <typename TIn, typename TOut>
+TAmounts<TIn, TOut>
+AMMLiquidity<TIn, TOut>::generateFibSeqOffer(
+    TAmounts<TIn, TOut> const& balances) const
+{
+    auto const n = ammContext_.curIters();
+    TAmounts<TIn, TOut> cur{};
+    Number x{};
+    Number y{};
+
+    cur.in = toAmount<TIn>(
+        getIssue(balances.in),
+        InitialFibSeqPct * initialBalances_.in,
+        Number::rounding_mode::upward);
+    cur.out = swapAssetIn(initialBalances_, cur.in, tradingFee_);
+    y = cur.out;
+    if (n == 0)
+        return cur;
+
+    std::uint16_t i = 0;
```

Comment:
Essentially, Fibonacci calculation is about applying a matrix multiplication multiple times.

`[x, y] = [[0, 1], [1,1]] * [x, y]`

It is possible to optimize this matrix multiplication (e.g. [here](https://stackoverflow.com/questions/3337455/how-to-generate-fibonacci-faster) or [here](https://stackoverflow.com/questions/18172257/efficient-calculation-of-fibonacci-series)). It can also be done at compile time.

Reply:
Thanks. Since the number of AMM offers is limited, can just hard-code the sequence and get the requested nth offer directly.

Reply:
@drlongle My real concern with the Fibonacci is that it's like coming across a function like:
```
y = x*1.5123413
```
And I ask: "How was that `1.5123413` chosen? Why don't we use something simpler like x=2*x" and I never get responses justifying the "complex" function. It's useful to know there are fast ways of computing it, but that doesn't answer the underlying question.

Reply:
@seelabs I only suggested that it's possible to compute the Fibonacci sequence fast. I leave it to @gregtatcam to answer why Fibonacci is used instead of something simpler.